File: src/modules/Pieces/Kinds/Knight/Knight.cpp
#pragma once

#include "Knight.h"

Knight::Knight(PieceParams* params):
    Piece<RawPosition>(params)
{}

vector<RawPosition> Knight::getPossibleMoves() {
    RawPosition* rawPosition = this->pieceImpl->getRawPosition();

    int column = rawPosition->getColumn();
    int row    = rawPosition->getRow();

    vector<RawPosition> possibleMoves = vector<RawPosition>();

    possibleMoves.push_back(RawPosition(column - 1, row - 2));
    possibleMoves.push_back(RawPosition(column + 1, row - 2));
    possibleMoves.push_back(RawPosition(column + 2, row - 1));
    possibleMoves.push_back(RawPosition(column + 2, row + 1));
    possibleMoves.push_back(RawPosition(column + 1, row + 2));
    possibleMoves.push_back(RawPosition(column - 1, row + 2));
    possibleMoves.push_back(RawPosition(column - 2, row + 1));
    possibleMoves.push_back(RawPosition(column - 2, row - 1));

    return this->pieceImpl->filterValidMoves(possibleMoves);
}

File: packages/interactive_2d/chess/chess_react/src/Sketches/ChessSketch/elements/Pieces/Kinds/Knight.ts
import { Piece } from "../Piece";
import { Position } from "../types";

export class knight extends Piece {
    // TODO: Fucking horse thinks he can eat the king!!!

    /**
     * @override
     */
    public getPossibleMoves(): Position[] {
        const { column, row } = this.position;

        const possibleMoves = [
            { column : column - 1, row : row - 2 },
            { column : column + 1, row : row - 2 },
            { column : column + 2, row : row - 1 },
            { column : column + 2, row : row + 1 },
            { column : column + 1, row : row + 2 },
            { column : column - 1, row : row + 2 },
            { column : column - 2, row : row + 1 },
            { column : column - 2, row : row - 1 },
        ] as Position[];

        return this.filterValidMoves(possibleMoves);
    }
}
